File: backend/user-service/src/main/java/com/__01/APP/Tesis/controllers/UsuarioGeneralController.java
package com.__01.APP.Tesis.controllers;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.__01.APP.Tesis.dto.ApiResponse;
import com.__01.APP.Tesis.dto.LoginRequest;
import com.__01.APP.Tesis.dto.RegistroRequest;
import com.__01.APP.Tesis.dto.UsuarioGeneralResponse;
import com.__01.APP.Tesis.services.UsuarioGeneralService;

@RestController
@RequestMapping("/api/usuarios")
@CrossOrigin(origins = {"*"})
public class UsuarioGeneralController {

    private final UsuarioGeneralService usuarioGeneralService;

    public UsuarioGeneralController(UsuarioGeneralService usuarioGeneralService) {
        this.usuarioGeneralService = usuarioGeneralService;
    }

    @PostMapping("/registro")
    public ResponseEntity<ApiResponse<UsuarioGeneralResponse>> registro(@RequestBody RegistroRequest request) {
        try {
            UsuarioGeneralResponse response = usuarioGeneralService.registrar(
                request.getNombreUsuario(),
                request.getEmail(),
                request.getContrasena()
            );
            return ResponseEntity.status(HttpStatus.CREATED)
                .body(new ApiResponse<>(true, "Usuario registrado exitosamente", response));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(new ApiResponse<>(false, e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ApiResponse<>(false, "Error interno del servidor"));
        }
    }

    @PostMapping("/login")
    public ResponseEntity<ApiResponse<UsuarioGeneralResponse>> login(@RequestBody LoginRequest request) {
        try {
            UsuarioGeneralResponse usuario = usuarioGeneralService.verificarCredenciales(
                request.getNombreUsuario(),
                request.getContrasena()
            );
            if (usuario != null) {
                return ResponseEntity.ok(
                    new ApiResponse<>(true, "Login exitoso", usuario)
                );
            } else {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new ApiResponse<>(false, "Nombre de usuario o contraseña incorrectos"));
            }
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ApiResponse<>(false, "Error interno del servidor"));
        }
    }

    @GetMapping("/obtener/{id}")
    public ResponseEntity<ApiResponse<UsuarioGeneralResponse>> obtenerPorId(@PathVariable Long id) {
        try {
            UsuarioGeneralResponse usuario = usuarioGeneralService.obtenerPorId(id);
            if (usuario != null) {
                return ResponseEntity.ok(
                    new ApiResponse<>(true, "Usuario obtenido", usuario)
                );
            } else {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(new ApiResponse<>(false, "Usuario no encontrado"));
            }
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ApiResponse<>(false, "Error interno del servidor"));
        }
    }

    @GetMapping("/email/{email}")
    public ResponseEntity<ApiResponse<UsuarioGeneralResponse>> obtenerPorEmail(@PathVariable String email) {
        try {
            UsuarioGeneralResponse usuario = usuarioGeneralService.obtenerPorEmail(email);
            if (usuario != null) {
                return ResponseEntity.ok(
                    new ApiResponse<>(true, "Usuario obtenido", usuario)
                );
            } else {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(new ApiResponse<>(false, "Usuario no encontrado"));
            }
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ApiResponse<>(false, "Error interno del servidor"));
        }
    }

    @GetMapping("/listar")
    public ResponseEntity<ApiResponse<List<UsuarioGeneralResponse>>> listarTodos() {
        try {
            List<UsuarioGeneralResponse> usuarios = usuarioGeneralService.listar();
            return ResponseEntity.ok(
                new ApiResponse<>(true, "Usuarios obtenidos", usuarios)
            );
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ApiResponse<>(false, "Error interno del servidor"));
        }
    }

    @DeleteMapping("/delete/{id}")
    public ResponseEntity<ApiResponse<Void>> eliminar(@PathVariable Long id) {
        try {
            boolean eliminado = usuarioGeneralService.eliminar(id);
            if (eliminado) {
                return ResponseEntity.ok(
                    new ApiResponse<>(true, "Usuario eliminado exitosamente")
                );
            } else {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(new ApiResponse<>(false, "Usuario no encontrado"));
            }
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(new ApiResponse<>(false, "Error interno del servidor"));
        }
    }
}

File: backend/user-service/src/main/kotlin/com/__01/APP/Tesis/controllers/UsuarioGeneralController.kt
package com.__01.APP.Tesis.controllers

import com.__01.APP.Tesis.dto.ApiResponse
import com.__01.APP.Tesis.dto.LoginRequest
import com.__01.APP.Tesis.dto.RegistroRequest
import com.__01.APP.Tesis.dto.UsuarioGeneralResponse
import com.__01.APP.Tesis.services.UsuarioGeneralService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime

@RestController
@RequestMapping("/api/usuarios")
@CrossOrigin(origins = ["*"])
class UsuarioGeneralController(
    private val usuarioGeneralService: UsuarioGeneralService
) {

    @PostMapping("/registro")
    fun registro(@RequestBody request: RegistroRequest): ResponseEntity<ApiResponse<UsuarioGeneralResponse>> {
        return try {
            val usuario = usuarioGeneralService.registrar(
                request.nombreUsuario,
                request.email,
                request.contraseña
            )
            val response = usuario.toResponse()
            ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse(true, "Usuario registrado exitosamente", response))
        } catch (e: IllegalArgumentException) {
            ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(ApiResponse(false, e.message ?: "Error en el registro"))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse(false, "Error interno del servidor"))
        }
    }

    @PostMapping("/login")
    fun login(@RequestBody request: LoginRequest): ResponseEntity<ApiResponse<UsuarioGeneralResponse>> {
        return try {
            val usuario = usuarioGeneralService.verificarCredenciales(
                request.nombreUsuario,
                request.contraseña
            )
            if (usuario != null) {
                val response = usuario.toResponse()
                ResponseEntity.ok(ApiResponse(true, "Login exitoso", response))
            } else {
                ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(ApiResponse(false, "Nombre de usuario o contraseña incorrectos"))
            }
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse(false, "Error interno del servidor"))
        }
    }

    @GetMapping("/{id}")
    fun obtenerPorId(@PathVariable id: Long): ResponseEntity<ApiResponse<UsuarioGeneralResponse>> {
        return try {
            val usuario = usuarioGeneralService.obtenerPorId(id)
            if (usuario != null) {
                val response = usuario.toResponse()
                ResponseEntity.ok(ApiResponse(true, "Usuario obtenido", response))
            } else {
                ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse(false, "Usuario no encontrado"))
            }
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse(false, "Error interno del servidor"))
        }
    }

    @GetMapping("/email/{email}")
    fun obtenerPorEmail(@PathVariable email: String): ResponseEntity<ApiResponse<UsuarioGeneralResponse>> {
        return try {
            val usuario = usuarioGeneralService.obtenerPorEmail(email)
            if (usuario != null) {
                val response = usuario.toResponse()
                ResponseEntity.ok(ApiResponse(true, "Usuario obtenido", response))
            } else {
                ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse(false, "Usuario no encontrado"))
            }
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse(false, "Error interno del servidor"))
        }
    }

    @GetMapping
    fun listarTodos(): ResponseEntity<ApiResponse<List<UsuarioGeneralResponse>>> {
        return try {
            val usuarios = usuarioGeneralService.listar()
            val response = usuarios.map { it.toResponse() }
            ResponseEntity.ok(ApiResponse(true, "Usuarios obtenidos", response))
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse(false, "Error interno del servidor"))
        }
    }

    @DeleteMapping("/{id}")
    fun eliminar(@PathVariable id: Long): ResponseEntity<ApiResponse<Nothing>> {
        return try {
            val eliminado = usuarioGeneralService.eliminar(id)
            if (eliminado) {
                ResponseEntity.ok(ApiResponse(true, "Usuario eliminado exitosamente"))
            } else {
                ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(ApiResponse(false, "Usuario no encontrado"))
            }
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse(false, "Error interno del servidor"))
        }
    }
}

private fun com.__01.APP.Tesis.models.UsuarioGeneral.toResponse() = UsuarioGeneralResponse(
    id = id,
    nombreUsuario = nombreUsuario,
    email = email,
    activo = activo,
    creadoEn = creadoEn,
    actualizadoEn = actualizadoEn
)
